server/integration/endpoint/src/main/java/org/infinispan/server/endpoint/subsystem/MultiTenantRouterService.java

```diff
@@ -0,0 +1,127 @@
+ * MA  02110-1301, USA.
+ */
+package org.infinispan.server.endpoint.subsystem;
+
+import org.infinispan.server.hotrod.HotRodServer;
+import org.infinispan.server.router.MultiTenantRouter;
+import org.infinispan.server.router.configuration.builder.MultiTenantRouterConfigurationBuilder;
+import org.infinispan.server.router.routes.impl.NettyHandlerRouteDestination;
+import org.infinispan.server.router.routes.Route;
+import org.infinispan.server.router.routes.impl.SniNettyRouteSource;
+import org.jboss.as.domain.management.SecurityRealm;
+import org.jboss.as.network.NetworkUtils;
+import org.jboss.as.network.SocketBinding;
+import org.jboss.msc.service.Service;
+import org.jboss.msc.service.StartContext;
+import org.jboss.msc.service.StartException;
+import org.jboss.msc.service.StopContext;
+import org.jboss.msc.value.InjectedValue;
+
+import java.net.InetSocketAddress;
+import java.util.HashMap;
+import java.util.Optional;
+
+import static org.infinispan.server.endpoint.EndpointLogger.ROOT_LOGGER;
+
+/**
+ * Multi tenant router service
+ *
+ * @author Sebastian Łaskawiec
+ */
+class MultiTenantRouterService implements Service<MultiTenantRouter> {
+
+    static class HotRodRouting {
+        private final InjectedValue<HotRodServer> hotRod = new InjectedValue<>();
+        private final InjectedValue<SecurityRealm> securityRealm = new InjectedValue<>();
+
+        public InjectedValue<HotRodServer> getHotRod() {
+            return hotRod;
+        }
+
+        public InjectedValue<SecurityRealm> getSecurityRealm() {
+            return securityRealm;
+        }
+    }
+
+    private static final String DEFAULT_NAME = "Multitenant Router";
+
+    private final InjectedValue<SocketBinding> socketBinding = new InjectedValue<>();
+    private final InjectedValue<HotRodServer> hotrod = new InjectedValue<>();
+    private final java.util.Map<String, HotRodRouting> sniRouting = new HashMap<>();
+
+    private final String name;
+    private final MultiTenantRouterConfigurationBuilder configurationBuilder;
+    private Optional<MultiTenantRouter> router = Optional.empty();
+
+    MultiTenantRouterService(MultiTenantRouterConfigurationBuilder configurationBuilder, Optional<String> serverName) {
+        this.name = constructServerName(serverName);
+        this.configurationBuilder = configurationBuilder;
+    }
+
+    private final String constructServerName(Optional<String> name) {
+        return name.orElse(DEFAULT_NAME);
+    }
+
+    @Override
+    public synchronized void start(final StartContext context) throws StartException {
+        ROOT_LOGGER.endpointStarting(name);
+        try {
+            SocketBinding socketBinding = getSocketBinding().getValue();
+            InetSocketAddress socketAddress = socketBinding.getSocketAddress();
+            configurationBuilder.ip(socketAddress.getAddress());
+            configurationBuilder.port(socketAddress.getPort());
+
+            sniRouting.entrySet().stream().forEach(e -> {
```

Comment:
TBH, I want us to avoid using `forEach` TBH since it's a side-effecting method. I'd much rather see the sniRouting entries be `map`'ed to a Stream of Routes and the add those to the routing itself. That would make the function pure and hence can be optimised better by the JVM rather than having lambdas like this that rely on side effects by capturing instance variables such as `configurationBuilder`.